sdk/src/actor.rs

```diff
@@ -107,6 +107,21 @@ pub fn create_actor(
     }
 }
 
+/// Upgrades an actor using the given block which includes the old code cid and the upgrade params
+pub fn upgrade_actor(new_code_cid: Cid, params: Option<IpldBlock>) -> SyscallResult<Response> {
```

Comment:
We can leave it like this for now, but we should have a better API.

- We generally take `Cid` by reference (it's large so we avoid copying till we need to).
- There is no "success" case here.

It should probably look something like:

```suggestion
pub enum UpgradeError {
    CodeNotInstalled, // code not available, could be worded better.
    InvalidUpgradeTarget, // code doesn't implement the upgrade entrypoint.
    ReentrentUpgrade, // actor already on the call stack.
    UpgradeFailed(Response),
}
pub fn upgrade_actor(new_code_cid: Cid, params: Option<IpldBlock>) -> Result<std::convert::Infallible, UpgradeError> {
```

I wish we could use `!` (the correct never type) but that's not stable yet.

Reply:
The reason to do this is that we want to treat _every_ possible return here as an error. If the user blindly writes `upgrade_actor(...)?` or `upgrade_actor(...).expect("upgrade failed")`, that should work.

Reply:
Hm. `ReentrentUpgrade` may not _quite_ work (see my other comments on the conflict with selfdestruct). Honestly, I'd just change the variant in this enum to `IllegalOperation` and document it as an "either or" case (where, 99% of the time, it'll be because the actor doesn't exist).

Reply:
I changed to  take CID by reference, will take a look at infallable/UpgradeError next

Reply:
Per our conversation below, `InvalidUpgradeTarget` likely _isn't_ something we need to include in this enum (that's covered by an exit code, not an error number, and we should probably leave the exit codes alone).